tfhe/src/integer/server_key/radix/shift.rs

```diff
@@ -117,50 +117,82 @@ impl ServerKey {
     /// sks.unchecked_scalar_right_shift_assign(&mut ct, shift);
     ///
     /// // Decrypt:
-    /// let dec: u64 = cks.decrypt(&ct);
+    /// let dec = cks.decrypt(&ct);
     /// assert_eq!(msg >> shift, dec);
     /// ```
     pub fn unchecked_scalar_right_shift_assign<PBSOrder: PBSOrderMarker>(
         &self,
         ct: &mut RadixCiphertext<PBSOrder>,
-        shift: usize,
+        shift: u64,
     ) {
-        let tmp = self.key.message_modulus.0 as f64;
-
-        //number of bits of message
-        let nb_bits = tmp.log2() as usize;
+        // see parallel implementation for a bit more details
 
-        // 2^u = 2^{p*q+r} = 2^{p*(q+1)}*2^{r-p}
-        let quotient = shift / nb_bits;
+        debug_assert!(ct.block_carries_are_empty());
+        debug_assert!(self.key.carry_modulus.0 >= self.key.message_modulus.0 / 2);
 
-        //p-r
-        let modified_remainder = nb_bits - (shift % nb_bits);
+        if shift == 0 {
+            return;
+        }
 
-        //if r == 0
-        if modified_remainder == nb_bits {
-            self.full_propagate(ct);
-            self.blockshift_right_assign(ct, quotient);
-        } else {
-            // B/2^u = (B*2^{p-r}) / (2^{p*(q+1)})
-            self.unchecked_scalar_left_shift_assign(ct, modified_remainder);
+        let num_bits_in_block = (self.key.message_modulus.0 as f64).log2() as u64;
+        let rotations = ((shift / num_bits_in_block) as usize).min(ct.blocks.len());
+        let shift_within_block = shift % num_bits_in_block;
+        let num_blocks = ct.blocks.len();
 
-            // We partially propagate in order to not lose information
-            self.partial_propagate(ct);
-            self.blockshift_right_assign(ct, 1_usize);
+        // rotate left as the blocks are from LSB to MSB
+        ct.blocks.rotate_left(rotations);
+        for block in &mut ct.blocks[num_blocks - rotations..] {
+            *block = self.key.create_trivial(0);
+        }
```

Comment:
I was wondering at some point, would be interesting to have a "fill_with_trivial(val: u64)" function on the blocks ?

Reply:
to avoid some allocations/assignments and re-use the existing memory

Reply:
Yes that would be cool, that would have much impact on performances, but would make the code look nicer i think

Reply:
actually we have a `create_trivial_assign` I should use that intead